devito/types/array.py

```diff
@@ -0,0 +1,200 @@
+
+                # Let UB be a function that rounds up a value `x` to the nearest
+                # multiple of the SIMD vector length
+                vl = configuration['platform'].simd_items_per_reg(self.dtype)
+                ub = lambda x: int(ceil(x / vl)) * vl
+
+                fvd_halo_size = sum(self.halo[-1])
+                fvd_pad_size = (ub(fvd_halo_size) - fvd_halo_size) + vl
+
+                padding[-1] = (0, fvd_pad_size)
+            return tuple(padding)
+        elif isinstance(padding, int):
+            return tuple((0, padding) for _ in range(self.ndim))
+        elif isinstance(padding, tuple) and len(padding) == self.ndim:
+            return tuple((0, i) if isinstance(i, int) else i for i in padding)
+        else:
+            raise TypeError("`padding` must be int or %d-tuple of ints" % self.ndim)
+
+    @classmethod
+    def __dtype_setup__(cls, **kwargs):
+        return kwargs.get('dtype', np.float32)
+
+    @property
+    def _C_typename(self):
+        return ctypes_to_cstr(POINTER(dtype_to_ctype(self.dtype)))
+
+    @property
+    def scope(self):
+        return self._scope
+
+    @property
+    def sharing(self):
+        return self._sharing
+
+    @property
+    def _mem_stack(self):
+        return self._scope == 'stack'
+
+    @property
+    def _mem_heap(self):
+        return self._scope == 'heap'
+
+    @property
+    def _mem_local(self):
+        return self._sharing == 'local'
+
+    @property
+    def _mem_shared(self):
+        return self._sharing == 'shared'
+
+    @cached_property
+    def free_symbols(self):
+        return super().free_symbols - {d for d in self.dimensions if d.is_Default}
+
+    # Pickling support
+    _pickle_kwargs = AbstractFunction._pickle_kwargs + ['dimensions', 'scope', 'sharing']
+
+
+class PointerArray(ArrayBasic):
+
+    """
+    Symbol representing a pointer to an Array.
+
+    Parameters
+    ----------
+    name : str
+        Name of the symbol.
+    dimensions : Dimension
+        The pointer Dimension.
+    array : Array
+        The pointed Array.
+
+    Warnings
+    --------
+    PointerArrays are created and managed directly by Devito (IOW, they are not
+    expected to be used directly in user code).
+    """
+
+    is_PointerArray = True
+
+    def __new__(cls, *args, **kwargs):
+        kwargs.update({'options': {'evaluate': False}})
+        return AbstractFunction.__new__(cls, *args, **kwargs)
+
+    def __init_finalize__(self, *args, **kwargs):
+        super(PointerArray, self).__init_finalize__(*args, **kwargs)
+
+        self._array = kwargs['array']
+        assert self._array.is_Array
+
+    @classmethod
+    def __dtype_setup__(cls, **kwargs):
+        return kwargs['array'].dtype
+
+    @property
+    def _C_typename(self):
+        return ctypes_to_cstr(POINTER(POINTER(dtype_to_ctype(self.dtype))))
+
+    @property
+    def dim(self):
+        """Shortcut for self.dimensions[0]."""
+        return self.dimensions[0]
```

Comment:
So single dim even if it points to n-dim Array?

Reply:
yes. The "pointer dimension" is just one, ie in C it's like `int (*a)[..][..][..]..` , cannot be for example `int (**a)[..][..][..]`. This generalisation would have complicated the code, and we don't need it.

Basically, we need it for the "thread dimension"